README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# bigtime: Sparse Estimation of Large Time Series Models

<!-- badges: start -->

[![CRAN\_Version\_Badge](http://www.r-pkg.org/badges/version/bigtime)](https://cran.r-project.org/package=bigtime)
[![CRAN\_Downloads\_Badge](https://cranlogs.r-pkg.org/badges/grand-total/bigtime)](https://cran.r-project.org/package=bigtime)
[![License\_GPLv2\_Badge](https://img.shields.io/badge/License-GPLv2-yellow.svg)](https://www.gnu.org/licenses/old-licenses/gpl-2.0.html)
[![License\_GPLv3\_Badge](https://img.shields.io/badge/License-GPLv3-yellow.svg)](https://www.gnu.org/licenses/gpl-3.0.html)
<!-- badges: end -->

The goal of `bigtime` is to sparsely estimate large time series models
such as the Vector AutoRegressive (VAR) Model, the Vector AutoRegressive
with Exogenous Variables (VARX) Model, and the Vector AutoRegressive
Moving Average (VARMA) Model. The univariate cases are also supported.

## Installation

The package can be installed from CRAN using

``` r
install.packages("bigtime")
```

You can install the development version (= the latest version) of
bigtime from github as follows:

``` r
# install.packages("devtools")
devtools::install_github("ineswilms/bigtime")
```

## Plotting the data

<!-- We will use the time series contained in the `example` data set. The first ten columns in our dataset are used as endogenous time series in the VAR and VARMA models, and the  last five columns are used as exogenous time series in the VARX model. Note that we remove the last observation from our dataset as we will use this one to illustrate how to evaluate prediction performance. We start by making a plot of our data. -->

bigtime comes with example data sets created from a VAR, VARMA and VARX
DGP. These data sets are called `var.example`, `varma.example`, and
`varx.example` respectively and can be loaded into the environment by
calling `data(var.example)` and similarly for the others. We can have a
look at the `varx.example` data set by first loading it into the
environment and then, using a little utility function, plotting it.

``` r
library(bigtime)
suppressMessages(library(tidyverse)) # Will be used for nicer visualisations
#> Warning: package 'tidyverse' was built under R version 4.0.5
#> Warning: package 'ggplot2' was built under R version 4.0.3
#> Warning: package 'readr' was built under R version 4.0.4
#> Warning: package 'purrr' was built under R version 4.0.3
#> Warning: package 'dplyr' was built under R version 4.0.5
#> Warning: package 'stringr' was built under R version 4.0.3
#> Warning: package 'forcats' was built under R version 4.0.4
data(varx.example) # loading the varx example data



plot_series <- function(Y){
  as_tibble(Y) %>%
  mutate(Time = 1:n()) %>%
  pivot_longer(-Time, names_to = "Series", values_to = "vals") %>%
  mutate(Series = factor(Series, levels = colnames(Y))) %>%
  ggplot() +
  geom_line(aes(Time, vals)) + 
  facet_wrap(facets = vars(Series), ncol = 1) + 
  ylab("") +
  theme_bw()
}

plot_series(Y.varx)
```

![](man/figures/README-unnamed-chunk-1-1.png)<!-- -->

``` r
plot_series(X.varx)
```

![](man/figures/README-unnamed-chunk-1-2.png)<!-- -->

## Multivariate Time Series Models

### Vector AutoRegressive (VAR) Models

#### Simulation

While we could use the example data provided, *bigtime* also supports
simulation of VAR models using both lasso and elementwise type sparsity
patterns. Since lasso is the most widely known one, we will start off
with lasso. To simulate a VAR model having lasso type of sparsity, use
the `simVAR` function and set `sparsity_pattern="lasso"`. The lasso
sparsity pattern has the additional `num_zero` option which determines
the number of zeros in the VAR coefficient matrix (excluding
intercepts). *Note: we also set a seed so that the simulation is
replicable*. We can then use the `summary` function to obtain a summary
of the simulated data.

``` r
periods <- 200 # time series length
k <- 5 # number of time series
p <- 8 # maximum lag 
sparsity_options <- list(num_zero = 15) # 15 zeros across the k=5 VAR coefficient matrices
sim_data <- simVAR(periods = periods, k = k, p = p, 
                   sparsity_pattern = "lasso", 
                   sparsity_options = sparsity_options,
                   seed = 123456, 
                   decay = 0.01) # the smaller decay the larger early coefs w.r.t. to later ones
Y.lasso <- sim_data$Y
summary(sim_data) # obtaining a summary of the simulated time series data
#> #### General Information #### 
#> 
#> Seed                                        123456 
#> Time series length                          200 
#> Burnin                                      200 
#> Variables Simulated                         5 
#> Number of Lags                              8 
#> Coefficients were randomly created?         TRUE 
#> Maximum Eigenvalue of Companion Matrix      0.8 
#> Sparsity Pattern                            lasso 
#> 
#> 
#> #### Sparsity Options #### 
#> 
#> $num_zero
#> [1] 15
#> 
#> 
#> 
#> #### Coefficient Matrix #### 
#> 
#>                  Y1            Y2            Y3            Y4            Y5
#> Y1.L1  3.344178e-01 -1.107252e-01 -1.423944e-01  3.509198e-02  9.033998e-01
#> Y2.L1  3.347094e-01  5.264215e-01  1.003833e+00  4.685881e-01 -1.709388e-01
#> Y3.L1 -3.995565e-01 -4.468152e-01 -2.235442e-02  4.710753e-01  4.224553e-01
#> Y4.L1  2.310627e-02 -2.948322e-01  3.732432e-01  6.691342e-01  2.244958e-01
#> Y5.L1  0.000000e+00  5.040150e-01  1.543970e-02  7.602611e-02  1.855509e-01
#> Y1.L2 -1.714191e-03  6.652793e-05  2.827333e-03  3.898174e-03 -2.488847e-03
#> Y2.L2 -3.432958e-03  2.790046e-04 -4.196394e-03 -1.102596e-02 -4.531967e-03
#> Y3.L2 -3.456294e-03  6.257595e-03  4.071610e-03  4.187554e-03 -4.475995e-03
#> Y4.L2  0.000000e+00  3.882016e-03  6.860267e-04 -3.594939e-03  6.349122e-04
#> Y5.L2 -2.013360e-03  0.000000e+00 -4.561977e-04  4.355841e-03 -4.860029e-03
#> Y1.L3 -7.090490e-05 -1.972221e-05  1.289428e-05  0.000000e+00  0.000000e+00
#> Y2.L3 -1.362040e-05 -4.321743e-05 -5.979591e-05 -1.013789e-05 -4.890420e-06
#> Y3.L3 -2.603130e-05  1.255775e-05  4.926041e-06 -3.356641e-05  2.408345e-05
#> Y4.L3 -9.864658e-06 -7.407063e-06  9.288386e-07 -1.943981e-05 -2.959808e-05
#> Y5.L3  5.224473e-05  2.264257e-05 -7.240193e-05  1.758216e-05 -5.780336e-05
#> Y1.L4  3.821882e-07 -2.899947e-07  1.955819e-08 -6.271466e-07 -9.234996e-07
#> Y2.L4  4.644695e-07 -2.826756e-07 -6.312740e-07  2.079154e-07 -4.271531e-07
#> Y3.L4  1.887393e-08  3.401594e-07  1.735509e-07  2.097019e-07  0.000000e+00
#> Y4.L4 -1.992917e-07  5.054378e-07  2.266186e-07 -1.382372e-07  2.907276e-07
#> Y5.L4  3.465950e-07  1.481081e-07  6.351945e-07  2.421511e-08  5.162710e-08
#> Y1.L5 -3.412591e-09  4.406386e-09 -4.838797e-09  2.333901e-09  3.464337e-09
#> Y2.L5  3.943109e-09 -5.099138e-09 -4.262808e-12 -3.854296e-09  5.050800e-09
#> Y3.L5 -3.476037e-09  9.989383e-10 -3.456284e-10 -1.977003e-09 -1.078979e-09
#> Y4.L5 -1.601388e-09 -3.787661e-09  4.063988e-09 -1.639348e-09 -7.263552e-10
#> Y5.L5  4.748150e-09 -6.451814e-09 -9.114958e-09 -8.185059e-09 -2.599211e-09
#> Y1.L6  2.878848e-11 -4.147466e-12  0.000000e+00 -3.921230e-11  0.000000e+00
#> Y2.L6  1.757755e-12 -4.097641e-11  3.182509e-11  2.012060e-11 -6.235119e-11
#> Y3.L6  3.543788e-11 -8.333751e-12  7.345804e-11  7.110232e-12 -3.232298e-11
#> Y4.L6  7.813330e-11  4.861321e-12 -1.423732e-11  4.223140e-11  4.665353e-11
#> Y5.L6 -3.204189e-11  3.110420e-11 -8.002152e-11 -1.406136e-11  1.527682e-11
#> Y1.L7  1.035718e-13  3.868736e-13  0.000000e+00  3.274603e-14 -6.488767e-14
#> Y2.L7  5.105382e-13 -2.889064e-13  6.181295e-13  3.424112e-13  3.218496e-13
#> Y3.L7 -1.374288e-13  4.027135e-14  6.096383e-14  0.000000e+00 -1.371909e-13
#> Y4.L7 -5.274508e-13 -7.661388e-13  2.324513e-13 -6.879730e-13  0.000000e+00
#> Y5.L7 -6.332500e-13  7.399670e-13  3.913858e-13 -1.324464e-13  4.786837e-13
#> Y1.L8 -1.337775e-15  6.673161e-16  0.000000e+00  7.885634e-15  2.243677e-15
#> Y2.L8 -2.573135e-15 -1.884193e-16  4.549152e-15  1.160132e-15  1.854712e-15
#> Y3.L8  4.740580e-15  0.000000e+00 -2.501340e-15 -2.635646e-15  2.839144e-15
#> Y4.L8  0.000000e+00 -1.797436e-15  4.488466e-15  7.460492e-16 -1.452254e-15
#> Y5.L8  9.907469e-16  1.438249e-15  6.791980e-17 -9.437176e-16  0.000000e+00
```

![](man/figures/README-unnamed-chunk-2-1.png)<!-- -->

A VAR with HLag (elementwise) type of sparsity can be simulated using
`simVAR` and by setting `sparsity_pattern="HLag"`. Three extra options
exist: (1) `zero_min` determines the minimum number of zero coefficients
of each variable in each equation, (2) `zero_max` determines the maximum
number of zero coefficients of each variable in each equation, and (3)
`zeroes_in_self` is a boolean option that if TRUE, zero coefficients of
the *i*th variable can exist in the *i*th equation.

``` r
periods <- 200
k <- 5 # Number of time series
p <- 5 # Maximum lag 
sparsity_options <- list(zero_min = 0, 
                         zero_max = 5, 
                         zeroes_in_self = TRUE)
sim_data <- simVAR(periods = periods, k = k, p = p, 
                   sparsity_pattern = "HLag", 
                   sparsity_options = sparsity_options,
                   seed = 123456, 
                   decay = 0.01)
Y.hlag <- sim_data$Y
summary(sim_data) # Obtaining a summary of the simulation
#> #### General Information #### 
#> 
#> Seed                                        123456 
#> Time series length                          200 
#> Burnin                                      200 
#> Variables Simulated                         5 
#> Number of Lags                              5 
#> Coefficients were randomly created?         TRUE 
#> Maximum Eigenvalue of Companion Matrix      0.8 
#> Sparsity Pattern                            hvar 
#> 
#> 
#> #### Sparsity Options #### 
#> 
#> $zero_min
#> [1] 0
#> 
#> $zero_max
#> [1] 5
#> 
#> $zeroes_in_self
#> [1] TRUE
#> 
#> 
#> 
#> #### Coefficient Matrix #### 
#> 
#>                  Y1            Y2            Y3            Y4            Y5
#> Y1.L1  0.000000e+00 -1.032031e-01 -1.327209e-01  3.270802e-02  8.420276e-01
#> Y2.L1  3.119710e-01  4.906592e-01  9.356382e-01  4.367547e-01 -1.593261e-01
#> Y3.L1 -3.724128e-01 -4.164610e-01 -2.083578e-02  4.390729e-01  3.937560e-01
#> Y4.L1  2.153655e-02 -2.748029e-01  3.478871e-01  0.000000e+00  2.092447e-01
#> Y5.L1 -2.818808e-01  4.697750e-01  1.439081e-02  7.086130e-02  1.729456e-01
#> Y1.L2  0.000000e+00  0.000000e+00  2.635259e-03  3.633353e-03 -2.319768e-03
#> Y2.L2 -3.199742e-03  2.600505e-04  0.000000e+00 -1.027691e-02 -4.224090e-03
#> Y3.L2 -3.221492e-03  5.832487e-03  0.000000e+00  3.903074e-03 -4.171920e-03
#> Y4.L2  0.000000e+00  3.618292e-03  6.394217e-04  0.000000e+00  5.917797e-04
#> Y5.L2 -1.876583e-03 -3.611195e-03 -4.252061e-04  4.059929e-03 -4.529864e-03
#> Y1.L3  0.000000e+00  0.000000e+00  1.201831e-05  0.000000e+00  5.747130e-05
#> Y2.L3 -1.269510e-05 -4.028146e-05  0.000000e+00 -9.449180e-06 -4.558191e-06
#> Y3.L3 -2.426287e-05  1.170464e-05  0.000000e+00 -3.128609e-05  2.244735e-05
#> Y4.L3  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00 -2.758734e-05
#> Y5.L3  0.000000e+00  2.110436e-05 -6.748333e-05  1.638772e-05 -5.387651e-05
#> Y1.L4  0.000000e+00  0.000000e+00  1.822951e-08  0.000000e+00 -8.607620e-07
#> Y2.L4  4.329159e-07 -2.634722e-07  0.000000e+00  0.000000e+00 -3.981346e-07
#> Y3.L4  0.000000e+00  3.170507e-07  0.000000e+00  1.954558e-07 -9.491777e-08
#> Y4.L4  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
#> Y5.L4  0.000000e+00  1.380464e-07  5.920428e-07  0.000000e+00  4.811983e-08
#> Y1.L5  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  3.228988e-09
#> Y2.L5  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
#> Y3.L5  0.000000e+00  0.000000e+00  0.000000e+00 -1.842696e-09 -1.005678e-09
#> Y4.L5  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00
#> Y5.L5  0.000000e+00 -6.013512e-09 -8.495736e-09  0.000000e+00  0.000000e+00
```

![](man/figures/README-unnamed-chunk-3-1.png)<!-- -->

#### Lasso estimation

The above simulated data (and truly any other data) could then be
estimated using an L1-penalty (lasso penalty) on the autoregressive
coefficients. To this end, set the `VARpen` argument in the `sparseVAR`
function equal to L1. *Note: it is recommended to standardise the data.
Bigtime will give a warning if the data is not standardised but will not
stop you from continuing.* Setting `selection="none"`, the default,
allows us to specify the penalization we want. Furthermore, we can
predefine the maximum lag-order by changing the `p` argument to the
desired value. However, we do not recommend this, as the bigtime will,
by default, choose a maximum lag-order that is well suited in many
scenarios.

``` r
VAR.L1 <- sparseVAR(Y = scale(Y.lasso), # standardising the data
                    VARpen = "L1", # using lasso penalty
                    VARlseq = 1.5) # Specifying the penalty to be used. selection="none" is the default.
```

#### Tuning parameter selection

For the selection of the penalization parameter, we can either set the
`selection` argument to `"none"`, which would return a model for a
sequence of penalizations, or use time series cross-validation by
setting `selection="cv"`, or finally we could also use any of BIC, AIC,
or HQ information criteria by setting the `selection` arguments to any
of `"bic"`, `"aic"`, or `"hq"` respectively. We will start of by using
time series cross-validation and will therefore set `selection="cv"`.
The default is to use a cross-validation score based on one-step ahead
predictions but you can change the default forecast horizon under the
argument `h`.

``` r
VAR.L1 <- sparseVAR(Y=scale(Y.lasso),  # standardising the data
                   selection = "cv", # using time series cross-validation
                   VARpen = "L1") # using the lasso penalty
```

The `plot_cv` function can be used to investigate the cross-validation
procedure. The returned plot shows the mean MSFE for each penalization
together with error bars for plus-minus one standard deviation. The
black dashed line indicates the penalty parameter choice that lead to
the smallest MSFE in the CV procedure. The red dotted line, on the other
hand, shows the one-standard-error solution. It picks the most
parsimonious model within one standard error of the best
cross-validation score. The latter is the one that is chosen by default
in `sparseVAR`.

``` r
plot_cv(VAR.L1)
```

![](man/figures/README-unnamed-chunk-6-1.png)<!-- -->

Further investigation into the model can be done by using the function
`lagmatrix`, which returns the lagmatrix of the estimated autoregressive
coefficients. If entry (*i*, *j*) = *x*, this means that the sparse
estimator indicates the effect of time series *j* on time series *i* to
last for *x* periods. Setting the `returnplot` argument to `TRUE` will
return a heatmap for better visual inspection.

``` r
LhatL1 <- lagmatrix(fit=VAR.L1, returnplot=TRUE)
```

![](man/figures/README-unnamed-chunk-7-1.png)<!-- -->

The lag matrix is typically sparse as it contains some empty (i.e.,
zero) cells. However, VAR models estimated with a standard L1-penalty
are typically not easily interpretable as they select many high lag
order coefficients (i.e., large values in the lagmatrix).

To circumvent this problem, we advise using a lag-based hierarchically
sparse estimation procedure, which boils down to using the default
option HLag for the `VARpen` argument. This estimation procedure
encourages low maximum lag orders, often results in sparser lagmatrices,
and hence more interpretable models.

#### Hlag estimation

Models can be estimated using the hierarchical penalization by using the
default argument to `VARpen`, namely `HLag`. Model selection can again
be done by either setting `selection="none"` and obtaining a whole
sequence of models, or by using any of `cv, bic, aic, hq`.

``` r
VARHLag_none <- sparseVAR(Y=scale(Y.hlag), 
                          selection = "none") # HLag is the default VARpen option
VARHLag_cv <- sparseVAR(Y=scale(Y.hlag), 
                        selection = "cv")
VARHLag_bic <- sparseVAR(Y=scale(Y.hlag), 
                         selection = "bic") # This will also give a table for IC comparison showing the selected lambda for each IC
#> 
#> 
#> #### Selected the following lambda ####
#> 
#>      AIC      BIC       HQ 
#> 10.74023 10.74023 10.74023 
#> 
#> 
#> #### Details ####
#> 
#>      lambda         AIC         BIC          HQ
#> 1  138.7154     -0.9793     -0.9793     -0.9793
#> 2   83.1577     -1.7792     -1.6473     -1.7258
#> 3   49.8517     -3.0055     -2.7746      -2.912
#> 4   29.8853     -4.1453     -3.8155     -4.0118
#> 5   17.9158      -4.826     -4.4631     -4.6791
#> 6   10.7402 ==> -5.0769 ==> -4.6151   ==> -4.89
#> 7    6.4386     -5.0698     -4.3277     -4.7695
#> 8    3.8598     -4.6703     -3.0377     -4.0096
#> 9    2.3139     -2.7531      2.5737     -0.5974
#> 10   1.3872     -1.5512      6.5956      1.7457
```

#### Diagnostics

Depending on which selection procedure was used, various diagnostics can

NEWS.md
# bigtime 0.2.3
* package documentation adjusted due to breaking change in roxygen2 7.0.0 

# bigtime 0.2.2
* verbose arguments added to functions sparseVAR and sparseVARX
* correction lambda grid to hvaralgorithms
* removal SystemRequirements: C++11

# bigtime 0.2.1
* Dependency on Rcpp version >=1.0.7 added to the DESCRIPTION.

# bigtime 0.2.0

* Important starting note: Please update all R packages on which bigtime depends upon installing bigtime from CRAN. We have been informed on errors occurring when using older versions of the Rcpp package on which bigtime depends. In version 0.2.1 of bigtime (currently only available on github), we have therefore added the dependency on Rcpp version >=1.0.7
* `sparseVAR`, `sparseVARMA`, `sparseVARX` are all more efficient and faster now
* All estimation functions return corresponding S3 classes
* `fitted` and `residuals` functions were implemented for VAR, VARMA, VARX 
* `diagnostics_plot` was implemented for VAR, VARMA, VARX
* `plot_cv` was implemented; Allows to investigate the behavior in CV 
* `sparseVAR`, `sparseVARMA` and `sparseVARX` can now use information criteria to select the optimal penalization
* `simVAR` allow for the simulation of VAR models using different sparsity patterns; utility functions `summary` and `plot` were also implemented for simulated VARs
* Default selection procedure in `sparseVAR`, `sparseVARMA` and `sparseVARX` changed to "none". Will return a 3D array of estimations from this version on, unless a different selection procedure is chosen. **WARNING: This can break old code!**
* `recursiveforecast` was implemented for VAR models. 
* `is.stable` was implemented for VAR models
* in `plot_cv`, `directforecast`, and `lagmatrix` the model argument was removed. Functions know  automatically what model was used. **WARNING: This can break old code!** 
* All model functions will give warnings if the given data is not standardized
* added example data for VAR, VARMA, and VARX

# bigtime 0.1.0

* First release on CRAN


R/simVAR.R
#' Simulates a VAR(p) with various sparsity patterns
#'
#' @param periods Scalar indicating the desired time series length
#' @param k Number of time series
#' @param p Maximum lag number. In case of \code{sparsity_patter="none"} this will be
#' the actual number of lags for all variables
#' @param coef_mat Coefficient matrix in companion form. If not provided,
#' one will be simulated
#' @param const Constant term of VAR. Default is zero. Must be either a scalar,
#' in which case it will be broadcasted to a k-vector, or a k-vector
#' @param e_dist Either a function taking argument n indicating the number of
#' variables in the system, or a matrix of dimensions k x (periods+burnin)
#' @param init_y Initial values. Defaults to zero. Expects either a scalar or
#' a vector of length (k*p)
#' @param max_abs_eigval Maximum allowed eigenvalue of companion matrix.
#' Only applicable if coefficient matrix is being simulated
#' @param burnin Number of time points to be used for burnin
#' @param sparsity_pattern The sparsity pattern that should be simulated.
#' Options are: \code{"none"} for a dense VAR, \code{"lasso"} (or \code{"L1"})
#' for a VAR with random zeroes,
#' and \code{"hvar"} (or \code{"HLag"}) for an elementwise hierarchical sparsity pattern
#' @param sparsity_options Named list of additional options for
#' when sparsity pattern is lasso (L1) or hvar (HLag). For lasso (L1) the option \code{num_zero}
#' determines the number of zeros. For hvar (HLag), the options \code{zero_min} (\code{zero_max})
#' give the minimum (maximum) of zeroes for each variable in each equation,
#' and the option \code{zeroes_in_self} (boolean) determines if any of the
#' coefficients of a variable on itself should be zero.
#' @param decay How much smaller should parameters for later lags be. The
#' smaller, the larger will early parameters be w.r.t. later ones.
#' @param seed Seed to be used for the simulation
#' @param ... Additional arguments passed to \code{e_dist}
#' @export
#' @return Returns an object of S3 class \code{bigtime.simVAR} containing the following
#' \item{Y}{Simulated Data}
#' \item{periods}{Time series length}
#' \item{k}{Number of endogenous variables}
#' \item{p}{Maximum lag length; effective lag length might be shorter due to sparsity patterns}
#' \item{coef_mat}{Companion form of the coefficient matrix. Will be of
#' dimensions (\code{k}\code{p})x(\code{k}\code{p}). First \code{k} rows correspond
#' to the actual coefficient matrix.}
#' \item{is_coef_mat_simulated}{\code{TRUE} if the \code{coef_mat} was simulated, \code{FALSE} if
#' it was user provided}
#' \item{const}{Constant term}
#' \item{e_dist}{Errors used in the construction of the data}
#' \item{init_y}{Initial conditions}
#' \item{max_abs_eigval}{Maximum eigenvalue to which the companion matrix
#' was constraint}
#' \item{burnin}{Burnin period used}
#' \item{sparsity_pattern}{Sparsity pattern used}
#' \item{sparsity_options}{Extra options for the sparsity patterns used}
#' \item{seed}{Seed used for the simulation}
#' @examples
#' periods <- 200 # time series length
#' k <- 5 # number of variables
#' p <- 10 # maximum lag
#' sparsity_pattern <- "HLag" # HLag sparsity structure
#' sparsity_options <- list(zero_min = 0, # variables can be included with all lags
#'                          zero_max = 10, # but some could also include no lags
#'                          zeroes_in_self = TRUE)
#' sim <- simVAR(periods=periods, k=k, p=p, sparsity_pattern=sparsity_pattern,
#'               sparsity_options=sparsity_options, seed = 12345)
#' summary(sim)
simVAR <- function(periods, k, p, coef_mat = NULL, const = rep(0, k), e_dist = rnorm,
                   init_y = rep(0, k*p), max_abs_eigval = 0.8, burnin = periods,
                   sparsity_pattern = c("none", "lasso", "L1", "hvar", "HLag"),
                   sparsity_options = NULL, decay = 1/p,
                   seed = NULL,
                   ...){

  ##################################
  #### Check and Prep ##############
  ##################################

  if (is.null(seed)) warning("No seed is being used. Replication might be infeasible")
  if (!is.null(seed)) set.seed(seed)

  # Checking all variables
  if (periods <= 0) stop("Must simulate at least one period: periods > 0")
  if (k < 1) stop("Must have at least one variable: k >= 1")
  if (!is.function(e_dist) & !is.matrix(e_dist)) stop("e_dist must either be a function or a matrix of dimensions k x (periods + burnin)")
  if (is.matrix(e_dist)){
    if ((ncol(e_dist) < periods | nrow(e_dist) < k)) stop("e_dist must be a function or a matrix of dimensions k x (periods + burnin)")
  }
  if (!is.null(coef_mat)) {
    if (k*p != ncol(coef_mat)) stop("coef_mat must have k*p columns not ", ncol(coef_mat))
  }
  if (!(length(const) == 1 | length(const) == k)) stop("const must either be of length one or of length k, not of length", length(const))
  if (p < 1) stop("Must include at least one lag: p >= 1")
  if (!(length(init_y) == 1 | length(init_y) == k*p)) stop("init_y must be of length one or of length k*p, not of length", length(init_y))
  if (max_abs_eigval < 0 | max_abs_eigval >=1) stop("Maximum eigenvalue must be strictly between zero and one, not ", max_abs_eigval)
  if (burnin < 0) stop("burnin must be non-negative: burnin >= 0")
  if (burnin == 0) warning("Not burning in the series is not recommended")
  if (!is.null(coef_mat)) warning("Will not use provided sparsity_pattern since coef_mat was given")
  if (decay > 1 | decay <=0) stop("Decay parameter must be between 0 and 1 (including)")

  # Which sparsity pattern was chosen?
  sparsity_pattern <- match.arg(sparsity_pattern)
  if (sparsity_pattern == "L1") sparsity_pattern <- "lasso"
  if (sparsity_pattern == "HLag") sparsity_pattern <- "hvar"

  # Creating coef_mat if needed
  is_coef_mat_simulated <- is.null(coef_mat)
  if (is.null(coef_mat)) coef_mat <- create_rand_coef_mat(k, p, max_abs_eigval, sparsity_pattern, sparsity_options, decay = decay)

  # Getting error terms
  if (is.function(e_dist)) e_dist <- do.call(cbind, lapply(1:(periods+burnin), function(x) e_dist(n = k, ...)))
  e_dist <- rbind(e_dist, matrix(0, nrow = k*(p-1), ncol = ncol(e_dist)))

  # constructing the constent
  if (length(const) == 1) const <- rep(const, k)
  const <- c(const, rep(0, k*(p-1)))

  # constructing the intial values
  if (length(init_y) == 1) init_y <- rep(init_y, k*p)

  ######################################
  #### END Check and Prep ##############
  ######################################


  Y <- simVAR_cpp(periods, k, p, coef_mat, const, e_dist, init_y, burnin)
  colnames(Y) <- paste0("Y", 1:k)

  sim_data <- list(
    Y = Y,
    periods = periods,
    k = k,
    p = p,
    coef_mat = coef_mat,
    is_coef_mat_simulated = is_coef_mat_simulated,
    const = const,
    e_dist = e_dist,
    init_y = init_y,
    max_abs_eigval = max_abs_eigval,
    burnin = burnin,
    sparsity_pattern = sparsity_pattern,
    sparsity_options = sparsity_options,
    seed = seed
  )
  class(sim_data) <- "bigtime.simVAR"
  sim_data
}


#' Creates a random coefficient matrix
#'
#' @param k Number of time series
#' @param p Number of lags
#' @param max_abs_eigval if < 1, then the VAR will be stable
#' @param sparsity_pattern The sparsity pattern that should be simulated.
#' Options are: \code{"none"} for a dense VAR, \code{"lasso"} for a VAR with random zeroes,
#' and \code{"hvar"} for an elementwise hierarchical sparsity pattern
#' @param sparsity_options Named list of additional options for
#' when sparsity pattern is lasso or hvar. For lasso the option \code{num_zero}
#' determines the number of zeros. For hvar, the options \code{zero_min} (\code{zero_max})
#' give the minimum (maximum) of zeroes for each variable in each equation,
#' and the option \code{zeroes_in_self} (boolean) determines if any of the
#' coefficients of a variable on itself should be zero.
#' @param decay How fast should coefficients shrink when the lag increases.
#' @param ... Not currently used
#' @export
#' @return Returns a coefficient matrix in companion form of dimension \code{kp}x\code{kp}.
create_rand_coef_mat <- function(k, p,
                                 max_abs_eigval = 0.8,
                                 sparsity_pattern = c("none", "lasso", "hvar"),
                                 sparsity_options = NULL,
                                 decay = 0.5,
                                 ...){

  ##################################
  #### Check and Prep ##############
  ##################################


  if (max_abs_eigval <= 0) stop("max_abs_eigval must be strictly postive, max_abs_eigval > 0")
  if (k < 1) stop("Model must have at least one variable: k>=1")
  if (p < 1) stop("Model must contain at least one lag: p > 1")

  sparsity_pattern <- match.arg(sparsity_pattern)
  # phi_vals <- runif(k*p*k, min = -100, max = 100)
  phi_vals <- rnorm(k*p*k, sd = 10)
  phis <- matrix(phi_vals, nrow = k, ncol = p*k)
  scale <- decay^(seq(0, p-1, by = 1))
  scale <- rep(scale, each = k)
  phis <- t(apply(phis, 1, function(x) x*scale))

  if (sparsity_pattern == "lasso"){
    # Some phi_vals will randomly be set to zero
    num_zero <- sparsity_options$num_zero
    if(is.null(num_zero)) {
      warning("Number of coefficients to be zero (num_zero) was not specified in sparsity_options. Defaulting to ", k*p)
      num_zero <- k*p
    }
    set_zero <- sample(1:(k*k*p), num_zero, replace = FALSE)
    phis[set_zero] <- 0
  }
  else if (sparsity_pattern == "hvar"){
    zero_min <- sparsity_options$zero_min
    zero_max <- sparsity_options$zero_max
    zeroes_in_self <- sparsity_options$zeroes_in_self
    if (is.null(zero_min)){
      warning("Minimum of zero was not given in sparsity_options. Defaulting to 0")
      zero_min <- 0
    }
    if (is.null(zero_max)){
      warning("Maximum of zero was not given in sparsity_options. Defaulting to ", floor(p/2))
      zero_max <- floor(p/2)
    }
    if (is.null(zeroes_in_self)){
      warning("Not specified whether there should be zero in self lags. Defaulting to true")
      zeroes_in_self <- TRUE
    }
    if(zero_max < zero_min) stop("zero_min must be smaller-equal than zero_max")
    if (zero_min > p | zero_max > p) stop("zero_min and zero_max must be at most p")

    for (k1 in 1:k){
      for (k2 in 1:k){
        if (k2 == k1 & !zeroes_in_self) next
        zeros <- sample(zero_min:zero_max, 1)
        if (zeros == 0) next
        set_zero <- seq(k2, k*p, k)[(p+1-zeros):p]
        phis[k1, set_zero] <- 0
      }
    }
  }

  ######################################
  #### END Check and Prep ##############
  ######################################


  I <- diag(1, nrow = k*(p-1), k*(p-1))
  O <- matrix(0, nrow = k*(p-1), ncol = k)
  coef_mat <- rbind(phis, cbind(I, O))
  while (max(abs(eigen(coef_mat)$values))>max_abs_eigval) {
    phis <- phis*0.99
    coef_mat <- rbind(phis, cbind(I, O))
  }
  rownames(coef_mat) <- c(paste0("Y", 1:k), rep("", dim(coef_mat)[[1]]-k))
  colnames(coef_mat) <- paste0(rownames(coef_mat)[1:k], ".L", rep(1:p, each = k))

  coef_mat
}

#' Plots a simulated VAR
#'
#' @param x Simulated data of class \code{bigtime.simVAR} obtained
#' from the \code{\link{simVAR}} function
#' @param ... Not currently used
#' @export
#' @return Returns a ggplot2 plot
plot.bigtime.simVAR <- function(x, ...){
  Time <- SimData <- Series <- NULL # Needed because of non-standard evaluation

  sim_data <- x
  Y <- sim_data$Y
  # if (!require(tidyverse, quietly = TRUE)) stop("tidyverse must be installed")
  dplyr::as_tibble(Y) %>%
    dplyr::mutate(Time = 1:dplyr::n()) %>%
    tidyr::pivot_longer(-Time, names_to = "Series", values_to = "SimData") %>%
    ggplot2::ggplot() +
    ggplot2::geom_line(ggplot2::aes(Time, SimData)) +
    ggplot2::facet_grid(rows = dplyr::vars(Series)) +
    ggplot2::ylab("") +
    ggplot2::theme_bw()
}


#' Gives a small summary of a VAR simulation
#' @param object Simulated data of class \code{bigtime.simVAR} obtained
#' from the \code{\link{simVAR}} function
#' @param plot Should the VAR be plotted. Default is \code{TRUE}
#' @param ... Not currently used
#' @export
#' @return If `plot=TRUE`, then a ggplot2 plot will be returned
summary.bigtime.simVAR <- function(object, plot = TRUE, ...){
  sim_data <- object
  cat("#### General Information #### \n\n")
  cat("Seed                                       ", sim_data$seed, "\n")
  cat("Time series length                         ", sim_data$periods, "\n")
  cat("Burnin                                     ", sim_data$burnin, "\n")
  cat("Variables Simulated                        ", sim_data$k, "\n")
  cat("Number of Lags                             ", sim_data$p, "\n")
  cat("Coefficients were randomly created?        ", sim_data$is_coef_mat_simulated, "\n")
  cat("Maximum Eigenvalue of Companion Matrix     ", sim_data$max_abs_eigval, "\n")
  cat("Sparsity Pattern                           ", sim_data$sparsity_pattern, "\n")

  cat("\n\n#### Sparsity Options #### \n\n")
  print(sim_data$sparsity_options)

  cat("\n\n#### Coefficient Matrix #### \n\n")
  print(t(sim_data$coef_mat[1:sim_data$k, ]))

  if(plot) plot(sim_data)
}






